docs/Prodigy_training_data.md

```diff
@@ -145,53 +129,83 @@ A comparison with the previous results (201022 - old definition, 210128 - new de
+| 210221 | bert | naive_bayes | bert | 0.730 | 0.746 | 0.803 | 0.696 |
 | 201022 | bert | SVM | bert | 0.819 | 0.881 | 0.825 | 0.945 |
 | 210128 | bert | SVM | bert | 0.822 | 0.809 | 0.752 | 0.874 |
-| 210218 | bert | SVM | bert | 0.815| 0.785 | 0.804 | 0.767 |
+| 210221 | bert | SVM | bert | 0.803 | 0.780 | 0.815 | 0.748 |
 | 201022 | bert | log_reg | bert | 1.000 | 0.825 | 0.797 | 0.855 |
 | 210128 | bert | log_reg | bert | 1.000 | 0.775 | 0.758 | 0.793 |
-| 210218 | bert | log_reg | bert | 0.992| 0.824 | 0.822 | 0.827 |
+| 210221 | bert | log_reg | bert | 0.993 | 0.789 | 0.817 | 0.763|
 | 201022 | bert | naive_bayes | scibert | 0.772 | 0.796 | 0.811 | 0.782 |
 | 210128 | bert | naive_bayes | scibert | 0.816 | 0.842 | 0.802 | 0.885 |
-| 210218 | bert | naive_bayes | scibert | 0.776| 0.800 | 0.814 | 0.787 |
+| 210221 | bert | naive_bayes | scibert | 0.764 | 0.779 | 0.803 | 0.756 |
 | 201022 | bert | SVM | scibert | 0.776 | 0.879 | 0.904 | 0.855 |
 | 210128 | bert | SVM | scibert | 0.831 | 0.851 | 0.819 | 0.885 |
-| 210218 | bert | SVM | scibert | 0.813| 0.817 | 0.849 | 0.787 |
+| 210221 | bert | SVM | scibert | 0.793 | 0.783 | 0.839 | 0.733 |
 | 201022 | bert | log_reg | scibert | 1.000 | 0.814 | 0.762 | 0.873 |
 | 210128 | bert | log_reg | scibert | 1.000 | 0.775 | 0.758 | 0.793 |
-| 210218 | bert | log_reg | scibert | 0.997| 0.853 | 0.853 | 0.853 |
+| 210221 | bert | log_reg | scibert | 1.000 | 0.794 | 0.819 | 0.770 |
 
 ### Ensemble model
 
-I ran:
+In the previous method we ran the script `ensemble_model.py` to make predictions using an ensemble of models which satisfied precision and recall value thresholds. However in this iteration we improved the process by testing which ensemble of models and which number of models that needs to agree is optimal.
+
+We first make predictions using the test data for each of the 12 models. This is done by running:
+```
+python nutrition_labels/predict_tech.py 
+    --training_data_file data/processed/training_data/210221/training_data.csv
+    --grants_data_file data/raw/wellcome-grants-awarded-2005-2019.csv
+    --models_date 210221
+```
+
+The ensemble parameter experiments were run in `notebooks/Ensemble parameter exploration.ipynb`. In this we plot the precision/recall for each of the 12 models plus every single combination of those 12 models for different values of the number of models that need to agree.
+
+![](figures/model_performance_ensemble_vs_single.png)
+
+The models found with the 'best' recall and precision were:
+
+- Ensemble of 3 models with 2 needing to agree ('bert_log_reg_bert_210221', 'bert_SVM_bert_210221', 'count_naive_bayes_210221')
+- Ensemble of 4 models with 3 needing to agree ('bert_naive_bayes_bert_210221', 'bert_log_reg_bert_210221', 'bert_SVM_bert_210221', 'count_naive_bayes_210221')
+- Ensemble of 4 models with 3 needing to agree ('bert_log_reg_bert_210221', 'bert_SVM_bert_210221', 'bert_naive_bayes_scibert_210221', 'count_naive_bayes_210221')
```

Comment:
it is worth saying that this ensemble technique is called majority voting and you can use VotingClassifier to easily implement it https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.VotingClassifier.html

Reply:
additionally if you want to extend the voting classifier to specify a more strict majority voting i.e. instead of simple majority a preselected number of models, I would recommend adding the extension to WellcomeML. I have already extended VotingClassifier to receive pretrained models https://github.com/wellcometrust/WellcomeML/blob/master/wellcomeml/ml/voting_classifier.py, you can introduce further parameters to do what you want

Reply:
ah yes - I was under the impression that our vectorizers (not all being from sklearn) would cause problems with VotingClassifier, but I will revisit that.

Reply:
I think our vectorisers should work as we have implemented them to be sklearn compatible, if they are not we should fix that.